python/ray/serve/_private/metrics_utils.py

```diff
@@ -307,71 +307,179 @@ def aggregate_avg(
+            is a list of TimeStampedValue objects sorted by timestamp.
+        ttl: Optional time-to-live for values. If specified, values expire after
+            this duration and are removed from the total.
+
+    Returns:
+        A list of TimeStampedValue representing the instantaneous total at event times.
+        Between events, the total remains constant (step function).
     """
-    if window_s <= 0:
-        raise ValueError(f"window_s must be positive, got {window_s}")
-
-    if not t1 and not t2:
+    if not replicas_timeseries or not any(replicas_timeseries):
         return []
 
-    # Align windows so each output timestamp sits at the start of its window.
-    # start is snapped to window_s boundary for binning stability
-    earliest = min(x[0].timestamp for x in (t1, t2) if x)
-    start = earliest // window_s * window_s
-
-    b1 = _bucket_latest_by_window(t1, start, window_s)
-    b2 = _bucket_latest_by_window(t2, start, window_s)
-
-    windows = sorted(set(b1.keys()) | set(b2.keys()))
-
+    # Build delta events per replica
+    # events[t] = total delta at time t across all replicas
+    events = defaultdict(float)  # time -> delta
+
+    if ttl is not None:
+        # expiry bookkeeping: last value + last time per replica
+        last_time = [None] * len(replicas_timeseries)
+        last_val = [0.0] * len(replicas_timeseries)
+
+    for rid, series in enumerate(replicas_timeseries):
+        if not series:
+            continue
+
+        # initial delta
+        t0, v0 = series[0].timestamp, series[0].value
+        events[t0] += v0
+        prev_t, prev_v = t0, v0
+
+        if ttl is not None:
+            last_time[rid], last_val[rid] = t0, v0
+            events[t0 + ttl] += -v0  # schedule initial expiry
+
+        for point in series[1:]:
+            t, v = point.timestamp, point.value
+            dv = v - prev_v
+            if dv != 0.0:
+                events[t] += dv
+            if ttl is not None:
+                # cancel previous expiry by adding back, then schedule new expiry
+                # (net effect: move expiry to t+ttl with the then-current value)
+                events[prev_t + ttl] += +prev_v  # cancel old expiry
+                events[t + ttl] += -v  # schedule new expiry
+                last_time[rid], last_val[rid] = t, v
+            prev_t, prev_v = t, v
+
+    # Sweep events (coalesce identical timestamps)
+    timeline = sorted(events.items())  # [(t, delta_sum_at_t), ...]
     merged: List[TimeStampedValue] = []
-    for w in windows:
-        v = b1.get(w, 0.0) + b2.get(w, 0.0)
-        ts_start = start + w * window_s
-        merged.append(TimeStampedValue(timestamp=ts_start, value=v))
+    current_sum = 0.0
+
+    for t, delta in timeline:
+        if delta == 0.0:
+            continue
+        current_sum += delta
+        # Record instantaneous total right after applying deltas at t
+        if merged and merged[-1].timestamp == t:
```

Comment:
Shouldn't `merged[-1].timestamp == t` never happen? It would imply a non-unique `t` in `timeline`'s keys, which doesn't make sense because it's a `dict`

Reply:
You are right, good call out, in the new implementation with heaps this can still happen.